src/Microsoft.ComponentDetection.Detectors/pip/ISimplePyPiClient.cs

```diff
@@ -0,0 +1,310 @@
+    private async Task<Stream> GetAndCacheProjectFileAsync(Uri uri)
+    {
+        if (!this.checkedMaxEntriesVariable)
+        {
+            this.cachedProjectWheelFiles = this.InitializeNonDefaultMemoryCache(this.cachedProjectWheelFiles);
+        }
+
+        if (this.cachedProjectWheelFiles.TryGetValue(uri, out Stream result))
+        {
+            this.cacheTelemetry.NumProjectFileCacheHits++;
+            this.logger.LogDebug("Retrieved cached Python data from {Uri}", uri);
+            return result;
+        }
+
+        var response = await this.GetPypiResponseAsync(uri);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            this.logger.LogWarning("Http GET at {ReleaseUrl} failed with status code {ResponseStatusCode}", uri, response.StatusCode);
+            return new MemoryStream();
+        }
+
+        var responseContent = await response.Content.ReadAsStreamAsync();
+
+        // The `first - wins` response accepted into the cache. This might be different from the input if another caller wins the race.
+        return await this.cachedProjectWheelFiles.GetOrCreateAsync(uri, cacheEntry =>
+        {
+            cacheEntry.SlidingExpiration = TimeSpan.FromSeconds(CACHEINTERVALSECONDS); // This entry will expire after CACHEINTERVALSECONDS seconds from last use
+            cacheEntry.Size = 1; // Specify a size of 1 so a set number of entries can always be in the cache
+            return Task.FromResult(responseContent);
+        });
+    }
+
+    /// <summary>
+    /// Returns a cached response if it exists, otherwise returns the response from PyPi REST call.
+    /// The response from PyPi is automatically added to the cache.
+    /// </summary>
+    /// <param name="uri">The REST Uri to call.</param>
+    /// <param name="spec">The PipDependencySpecification for the project. </param>
+    /// <returns>The cached deserialized json object or a new result from Simple PyPi.</returns>
+    private async Task<SimplePypiProject> GetAndCacheSimpleProjectAsync(Uri uri, PipDependencySpecification spec)
+    {
+        var pythonProject = new SimplePypiProject();
+        if (!this.checkedMaxEntriesVariable)
+        {
+            this.cachedSimplePyPiProjects = this.InitializeNonDefaultMemoryCache(this.cachedSimplePyPiProjects);
+        }
+
+        if (this.cachedSimplePyPiProjects.TryGetValue(uri, out SimplePypiProject result))
+        {
+            this.cacheTelemetry.NumSimpleProjectCacheHits++;
+            this.logger.LogDebug("Retrieved cached Python data from {Uri}", uri);
+            return result;
+        }
+
+        var response = await this.RetryPypiRequestAsync(uri, spec);
+
+        var responseContent = await response.Content.ReadAsStringAsync();
+        if (string.IsNullOrEmpty(responseContent))
+        {
+            return pythonProject;
+        }
+
+        try
+        {
+            pythonProject = JsonConvert.DeserializeObject<SimplePypiProject>(responseContent);
```

Comment:
Is it possible to refactor the Retry logic so we can utilize the async features of System.Text.Json? HttpClient has extension methods ([`HttpClient.GetFromJsonAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.json.httpclientjsonextensions.getfromjsonasync)) to make a GET request and deserialize json asynchronously

Reply:
synced with Jamie. might be a little complicated to redo the retry logic due to the status codes and retry criteria